File: titan/src/bank.ts
import { Expression } from './expression';
import { Money } from './money';
import { Sum } from './sum';

export class Bank {
  reduce(source: Expression, to: string): Money {
    return source.reduce(this, to);
  }

  addRate(from: string, to: string, rate: number): void {}
}

File: Yu/Sources/Test-Driven-Development-By-Example/Bank.swift
public class Bank {
    public init() {}
    var rates: [Pair: Int] = [:]
    public func reduce(_ source: any Expression, _ to: String) -> Money {
        return source.reduce(bank: self, to)
    }

    public func rate(_ from: String, _ to: String) -> Int {
        if from == to { return 1 }
        return rates[Pair(from: from, to: to)] ?? 0
    }

    public func addRate(_ from: String, _ to: String, _ rate: Int) {
        rates[Pair(from: from, to: to)] = rate
    }
}

struct Pair: Hashable {
    var from, to: String
}
